src/Confluent.Kafka/Producer.cs

```diff
@@ -175,14 +231,21 @@ public void HandleDeliveryReport(Message message)
             }
         }
 
-        internal void Produce(
+        internal void ProduceImpl(
             string topic,
             byte[] val, int valOffset, int valLength,
             byte[] key, int keyOffset, int keyLength,
-            DateTime? timestamp,
-            Int32 partition, bool blockIfQueueFull,
+            Timestamp timestamp,
+            Partition partition, 
+            IEnumerable<KeyValuePair<string, byte[]>> headers,
+            bool blockIfQueueFull,
```

Comment:
this parameter only required by the obsoleted methods. after they are removed, we can refactor this (internal) method.